ros/angel_system_nodes/angel_system_nodes/dummy_multi_task_monitor.py

```diff
@@ -0,0 +1,360 @@
+PARAM_TASK_ERROR_TOPIC = "task_error_topic"
+PARAM_QUERY_TASK_GRAPH_TOPIC = "query_task_graph_topic"
+
+
+@dataclass
+class TaskStateInformation:
+    """
+    Stores relevant task state information for a single task.
+    """
+
+    task_title: str
+    n_steps: int
+
+    # Tracks the current/previous steps
+    # Current step is the most recently completed step the user has finished
+    # as predicted by the HMM.
+    # Previous step is the step before the current step.
+    # Values will be either None or the step string label.
+    current_step = None
+    current_step_id = 0  # HMM ID
+    previous_step = None
+    previous_step_id = 0  # HMM ID
+    # Track the step skips we previously notified the user about to avoid
+    # sending duplicate notifications for the same error.
+    previous_step_skip = None
+    current_step_skip = None
+    # Track the indices of steps we have emitted skip errors for in order
+    # to not repeatedly emit many errors for the same steps.
+    # This may be removed from if step progress regresses.
+    steps_skipped_cache: Set[int] = field(default_factory=set)
+    # Track the latest activity classification end time sent to the HMM
+    # Time is in floating-point seconds up to nanosecond precision.
+    latest_act_classification_end_time = None
+    # List of task steps
+    steps: List[str] = field(default_factory=list)
+
+    def __post_init__(self):
+        # Boolean vector the length of task steps
+        # that indicates which steps to report as having been completed.
+        self.steps_complete = np.zeros(self.n_steps, dtype=bool)
+        # Skipped steps are those steps at or before the current
+        # skipped (haven't happened yet).
+        self.steps_skipped = np.zeros(self.n_steps, dtype=bool)
+
+
+class DummyMultiTaskMonitor(Node):
+    """
+    ROS node that simulates a node that could monitor multiple tasks
+    simultaneously. The keyboard is used to advance between task steps.
+    """
+
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+        log = self.get_logger()
+
+        param_values = declare_and_get_parameters(
+            self,
+            [
+                (PARAM_CONFIG_FILE,),
+                (PARAM_TASK_STATE_TOPIC,),
+                (PARAM_QUERY_TASK_GRAPH_TOPIC,),
+            ],
+        )
+        self._config_file = param_values[PARAM_CONFIG_FILE]
+        self._task_state_topic = param_values[PARAM_TASK_STATE_TOPIC]
+        self._query_task_graph_topic = param_values[PARAM_QUERY_TASK_GRAPH_TOPIC]
+
+        # Load the task configurations
+        with open(self._config_file, "r") as f:
+            config = yaml.safe_load(f)
+
+        self._task_state_dict: Dict[int, TaskStateInformation] = {}
+        for t in config["tasks"]:
+            # Load the config for this task
+            with open(t["config_file"], "r") as f:
+                task_config = yaml.safe_load(f)
+
+            n_steps = task_config["steps"][-1]["id"]
```

Comment:
Using the last ID might not be reliable. Would it not be better to take the `len(...)` here?